File: java-basic/src/main/java/ch15/Test15.java
// Object 클래스 - clone() 사용법 I 
package ch15;
 
// clone()은 인스턴스를 복제할 때 호출하는 메서드이다.

public class Test15 {
  
  static class Score {
    String name;
    int kor;
    int eng;
    int math;
    int sum;
    float aver;
    
    public Score() {}
    
    public Score(String name, int kor, int eng, int math) {
      this.name = name;
      this.kor = kor;
      this.eng = eng;
      this.math = math;
      this.sum = this.kor + this.eng + this.math;
      this.aver = this.sum / 3f;
    }

    @Override
    public String toString() {
      return "Score [name=" + name + ", kor=" + kor + ", eng=" + eng + ", math=" + math + ", sum="
          + sum + ", aver=" + aver + "]";
    }
  }
  
  public static void main(String[] args) {
    Score s1 = new Score("홍길동", 100, 100, 100);
    System.out.println(s1);
    
    // 인스턴스 복제
    // Object에서 상속 받은 clone()은 protected 접근 제한이 있다. 
    // 따라서 그냥 사용할 수 없다.
    //Score s2 = s1.clone(); // 컴파일 오류!
    
    // 해결책1:
    // => 직접 복제하라!
    Score s2 = new Score(s1.name, s1.kor, s1.eng, s1.math); // 빈 객체 생성
    s2.name = "임꺽정";
    
    // s1과 s2는 서로 다른 인스턴스이다.
    System.out.println(s1);
    System.out.println(s2);
    
    // 해결책2:
    // => Object에서 상속 받은 clone()을 오버라이딩 하라!
    // => Test16.java 를 살펴보라!
  }
}

File: node-basic/step01/test15.js
/* 주제: 배열
- 한 번에 여러 개의 메모리를 준비하는 방법
- 변수명과 인덱스 번호를 사용하여 메모리의 위치를 제어한다.
- 인덱스는 0부터 시작한다.
- 여러 개의 값을 다루기 편하다. 특히 반복문을 사용해 값을 다루기 편하다.
- 문법
var names = [];  <--- 빈 배열 생성
names[0] = "홍길동";
*/
"use strict"
// 배열 사용 전
var kor = 100, eng = 100, math = 100;
var sum = kor + eng + math;
var aver = sum / 3;
console.log(kor, eng, math, sum, aver);

// 배열 사용 후
var score = []; // 배열을 사용하기 전에 반드시 빈 배열을 만들어야 한다.
score[0] = 90;
score[1] = 80;
score[2] = 70
score[3] = score[0] + score[1] + score[2];
score[4] = score[3] / 3;
console.log(score[0], score[1], score[2], score[3], score[4]);
//
